### PR TITLE
refactor!: camelcase for exposed API

### PR DESCRIPTION
**Pull request by zhangyouxin**
This pr has:
- Refactor user interfaces to camelCase
  - Changed types in `api.d.ts` in `@ckb-lumos/base` to camelCase
  - Changed all corresponding types according to changes in previous step
  - Changed types used in `@ckb-lumos/ckb-indexer` to camelCase
  - Added param formatter and result formatter in `@ckb-lumos/ckb-indexer` to support camelCase
  - Fixed all failing tests caused by previous changes
- Removed old serilization method
  - Removed `core.js` in `@ckb-lumos/base`
  - Removed `Normalizers` and `Transformers` in `@ckb-lumos/toolkit`
  - Implemented new serilization methods in `@ckb-lumos/codec/blockchain.ts`
- Replaced `@ckb-lumos/rpc` with rpc module in https://github.com/nervosnetwork/ckb-sdk-js

### CODE REVIEW

File: examples/cardano-lock-namiwallet/lib.ts [hunk 1]
```diff
@@ -132,25 +133,25 @@ export async function transfer(options: Options): Promise<string> {
...
   let toSign = builder.make_data_to_sign().to_bytes();
 
-  const placeHolder = new toolkit.Reader(
+  const placeHolder = (
     "0x" +
       "00".repeat(
         SerializeCardanoWitnessLock({
-          pubkey: new toolkit.Reader("0x" + "00".repeat(32)).toArrayBuffer(),
-          signature: new toolkit.Reader("0x" + "00".repeat(64)).toArrayBuffer(),
+          pubkey: bytes.bytify("0x" + "00".repeat(32)),
```
Comment (homura) [inline examples/cardano-lock-namiwallet/lib.ts]: This can be refactored to `new Uint8Array(32)` or we can add a new function for bytes utils like this 

```ts
function create(length: number): Uint8Array;
```

File: packages/rpc/src/Base/index.ts [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+export class Base {
+  #rpcProperties: RpcPropertes = rpcProperties;
+
+  get rpcProperties(): RpcPropertes {
+    return this.#rpcProperties;
+  }
+}
+
+export default Base;
```
Comment (homura) [inline packages/rpc/src/Base/index.ts]: maybe we dont need the `export default` since modules are named export in Lumos

File: examples/cardano-lock-namiwallet/lib.ts [hunk 3]
```diff
@@ -32,8 +33,8 @@ config.initializeConfig(CONFIG);
 
 const CKB_RPC_URL = "https://testnet.ckb.dev/rpc";
 const CKB_INDEXER_URL = "https://testnet.ckb.dev/indexer";
-const rpc = new RPC(CKB_RPC_URL);
-const indexer = new Indexer(CKB_INDEXER_URL, CKB_RPC_URL);
+const rpc = new RPC.default(CKB_RPC_URL);
```
Comment (homura) [inline examples/cardano-lock-namiwallet/lib.ts]: Lumos's module are named export, we should keep the export style

File: examples/omni-lock-metamask/lib.ts [hunk 4]
```diff
@@ -94,51 +95,49 @@ export async function transfer(options: Options): Promise<string> {
...
   const messageForSigning = (() => {
     const hasher = new utils.CKBHasher();
 
-    const SECP_SIGNATURE_PLACEHOLDER = new toolkit.Reader(
+    const SECP_SIGNATURE_PLACEHOLDER = (
       "0x" +
         "00".repeat(
           SerializeRcLockWitnessLock({
-            signature: new toolkit.Reader("0x" + "00".repeat(65)),
+            signature: bytes.bytify("0x" + "00".repeat(65)),
```
Comment (homura) [inline examples/omni-lock-metamask/lib.ts]: can be refactored to `new Uint8Array(65)`

File: examples/secp256k1-multisig-transfer/index.ts [hunk 5]
```diff
@@ -154,23 +153,21 @@ export async function transfer(options: Options): Promise<string> {
     };
 
     if (witness !== "0x") {
-      const witnessArgs = new core.WitnessArgs(new toolkit.Reader(witness));
-      const lock = witnessArgs.getLock();
-      if (lock.hasValue() && new toolkit.Reader(lock.value().raw()).serializeJson() !== newWitnessArgs.lock) {
+      const witnessArgs = blockchain.WitnessArgs.unpack(bytes.bytify(witness))
+      const lock = witnessArgs.lock;
+      if (!!lock && lock !== newWitnessArgs.lock) {
```
Comment (homura) [inline examples/secp256k1-multisig-transfer/index.ts:158]: if the `witness.lock` is an object, comparing via `!==` will cause a bug

File: packages/base/lib/utils.d.ts [hunk 6]
```diff
@@ -1,68 +1,23 @@
-import { Reader } from "@ckb-lumos/toolkit";
-import { HexNumber } from "..";
-import { BI, BIish } from "@ckb-lumos/bi";
-import { Script, Input } from "./api";
-import { Hash, HexString } from "./primitive";
+import { Script } from "./api";
```
Comment (homura) [inline packages/base/lib/utils.d.ts]: why `utils.d.ts` has changed so much? It looks like an auto-generated file

File: packages/base/lib/utils.js [hunk 7]
```diff
@@ -151,20 +136,53 @@ function generateTypeIdArgs(input, outputIndex) {
...
+  if (Object.prototype.toString.call(data) === "[object Object]") {
+    for (let key in data) {
+      const value = data[key];
+      if (
+        Object.prototype.toString.call(value) === "[object Object]" ||
+        Object.prototype.toString.call(value) === "[object Array]"
+      ) {
+        result[toCamel(key)] = deepCamel(value);
+      } else {
+        result[toCamel(key)] = value === "dep_group" ? "depGroup" : value;
```
Comment (homura) [inline packages/base/lib/utils.js]: why we need to transform dep_group here? the `deepCamel` function name doesn't look honest

File: packages/codec/src/blockchainUtils.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  typeof blockchain.RawTransaction
+>;
+declare type CellDepCodecType = UnpackResult<typeof blockchain.CellDep>;
+declare type OutPointCodecType = UnpackResult<typeof blockchain.OutPoint>;
+declare type CellInputCodecType = UnpackResult<typeof blockchain.CellInput>;
+declare type BlockCodecType = UnpackResult<typeof blockchain.Block>;
+declare type UncleBlockCodecType = UnpackResult<typeof blockchain.UncleBlock>;
+declare type HeaderCodecType = UnpackResult<typeof blockchain.Header>;
+
+/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types */
```
Comment (homura) [inline packages/codec/src/blockchainUtils.ts]: we should try to avoid any

File: packages/ckb-indexer/README.md [hunk 9]
```diff
@@ -1,446 +1,279 @@
-CKB indexer is based on  [ckb-indexer](https://github.com/nervosnetwork/ckb-indexer) with more features. It is designed for:
+## Introduce
 
-- Web client usage.
-- CKB's RPC query.
+All the [indexing](https://github.com/nervosnetwork/ckb/tree/develop/rpc#indexer) RPCs built in the CKB has been deprecated since v0.36.0 and removed in v0.40.0. This is a standalone service for creating cell and transaction indexes as an alternate solution.
```
Comment (homura) [inline packages/ckb-indexer/README.md]: I dont quite understand why we need to refactor this document to re-introduce ckb-indexer's RPC, maybe we can just link to the README of ckb-indexer, right? 
A redundant document may lead to errors due to untimely updates

File: packages/ckb-indexer/src/indexer.ts [hunk 10]
```diff
@@ -38,46 +39,52 @@ function defaultLogger(level: string, message: string) {
...
+  public async tip(): Promise<IndexerType.Tip> {
+    return await this.getIndexerRpc().get_tip();
   }
 
   asyncSleep(timeout: number): Promise<void> {
     return new Promise((resolve) => setTimeout(resolve, timeout));
   }
 
   async waitForSync(blockDifference = 0): Promise<void> {
+    console.log(blockDifference);
```
Comment (homura) [inline packages/ckb-indexer/src/indexer.ts]: maybe we should remove conosle.log

File: packages/ckb-indexer/src/indexer.ts [hunk 11]
```diff
@@ -212,11 +217,21 @@ export class CkbIndexer implements Indexer {
...
     if (queries.lock) {
-      validators.ValidateScript(queries.lock);
-      emitter.lock = queries.lock as Script;
+      if (!instanceOfScriptWrapper(queries.lock)) {
+        validators.ValidateScript(queries.lock);
+        emitter.type = queries.lock;
+      } else if (instanceOfScriptWrapper(queries.lock)) {
+        validators.ValidateScript(queries.lock.script);
+        emitter.type = queries.lock.script;
+      }
```
Comment (homura) [inline packages/ckb-indexer/src/indexer.ts:224]: would you please add a test case for this part?

File: packages/ckb-indexer/src/rpc.ts [hunk 12]
```diff
@@ -1,78 +1,77 @@
...
+    return utils.deepCamel(await request(this.uri, "get_transactions", params));
   }
   async get_indexer_info(): Promise<string> {
-    return this.rpcProxy.get_indexer_info();
+    return utils.deepCamel(await request(this.uri, "get_indexer_info"));
   }
 }
+/* eslint-enable @typescript-eslint/no-explicit-any */
+
+/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types */
```
Comment (homura) [inline packages/ckb-indexer/src/rpc.ts:48]: Why we need `eslint-enable @typescript-eslint/no-explicit-any` twice here?

File: packages/ckb-indexer/src/rpcType.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
+import { HexString, HexNumber, Hash } from "@ckb-lumos/base";
+
+// eslint-disable-next-line @typescript-eslint/no-namespace
+export namespace RPCType {
```
Comment (homura) [inline packages/ckb-indexer/src/rpcType.ts:4]: We don't need the namespace here

```ts
import type RpcType from './rpcType'
```

File: packages/ckb-indexer/tests/cell_collector.unit.test.ts [hunk 14]
```diff
@@ -85,38 +80,6 @@ test("validateQueryOption#should throw error if lock and type not provided", (t)
   );
 });
 
-test("validateQueryOption#validate lock if lock is script", (t) => {
```
Comment (homura) [inline packages/ckb-indexer/tests/cell_collector.unit.test.ts:88]: any reason to remove these tests?

File: packages/ckb-indexer/tests/start_mock_rpc.ts [hunk 15]
```diff
@@ -2,10 +2,17 @@ import { createCKBMockRPC, mockData } from "@ckb-lumos/testkit";
...
 import path from "path";
 
+const blockDataPath = path.join(__dirname, "./blocks_data.json");
+const blocksData = JSON.parse(
+  fs.readFileSync(blockDataPath, "utf8").toString()
+);
+
+console.log("blockDataPath", blockDataPath);
+console.log("blocksData length", blocksData.length);
+console.log("blocksData[0].header", blocksData[0]["header"]);
```
Comment (homura) [inline packages/ckb-indexer/tests/start_mock_rpc.ts]: no console here

File: packages/codec/package.json [hunk 16]
```diff
@@ -30,7 +30,8 @@
     "release": "npm publish"
   },
   "dependencies": {
-    "@ckb-lumos/bi": "0.18.0"
+    "@ckb-lumos/bi": "0.18.0",
+    "@ckb-lumos/toolkit": "0.18.0"
```
Comment (homura) [inline packages/codec/package.json:34]: why we need toolkit in codec? the codec should be an independent module

File: packages/codec/src/molecule/layout.ts [hunk 17]
```diff
@@ -282,7 +282,7 @@ export function union<T extends Record<string, BytesCodec>>(
       const type = obj.type;
       const fieldIndex = fields.indexOf(type);
       if (fieldIndex === -1) {
-        throw new Error(`Unknown union type: ${obj.type}`);
+        throw new Error(`Unknown union type: ${String(obj.type)}`);
```
Comment (homura) [inline packages/codec/src/molecule/layout.ts:300]: Any reason for `String` wrapper? Is there any change here?

File: packages/common-scripts/src/anyone_can_pay.ts [hunk 18]
```diff
@@ -168,34 +168,23 @@ export async function setupInputCell(
...
     if (witness !== "0x") {
-      const witnessArgs = new core.WitnessArgs(new Reader(witness));
-      const lock = witnessArgs.getLock();
-      if (
-        lock.hasValue() &&
-        new Reader(lock.value().raw()).serializeJson() !== newWitnessArgs.lock
-      ) {
+      const witnessArgs = blockchain.WitnessArgs.unpack(bytes.bytify(witness));
+      const lock = witnessArgs.lock;
+      if (!!lock && lock !== newWitnessArgs.lock) {
```
Comment (homura) [inline packages/common-scripts/src/anyone_can_pay.ts:174]: maybe we should implement a `compare` function in codec/bytes to check if two `BytesLike` are equaling

File: packages/transaction-manager/package.json [hunk 19]
```diff
@@ -31,6 +31,7 @@
     "url": "https://github.com/nervosnetwork/lumos/issues"
   },
   "dependencies": {
+    "@ckb-lumos/rpc": "0.18.0",
```
Comment (homura) [inline packages/transaction-manager/package.json:34]: any reason to add the rpc as dependency here?

File: packages/base/lib/indexer.js [hunk 20]
```diff
@@ -1,6 +1,5 @@
-const { validators } = require("@ckb-lumos/toolkit");
 const utils = require("./utils");
-
+const { validators } = require("@ckb-lumos/toolkit");
```
Comment (PainterPuppets) [inline packages/base/lib/indexer.js:1]: maybe an unnecessary change?

File: packages/ckb-indexer/tests/start_mock_rpc.ts [hunk 21]
```diff
@@ -2,10 +2,17 @@ import { createCKBMockRPC, mockData } from "@ckb-lumos/testkit";
 import fs from "fs";
 import path from "path";
 
+const blockDataPath = path.join(__dirname, "./blocks_data.json");
+const blocksData = JSON.parse(
+  fs.readFileSync(blockDataPath, "utf8").toString()
+);
+
+console.log("blockDataPath", blockDataPath);
```
Comment (PainterPuppets) [inline packages/ckb-indexer/tests/start_mock_rpc.ts]: maybe need remove console

File: packages/codec/src/blockchain.ts [hunk 22]
```diff
@@ -86,6 +91,192 @@ const HexifyCodec = createBytesCodec<string, BytesLike>({
...
+
+export const CellDep = struct(
+  {
+    outPoint: OutPoint,
+    depType: DepType,
+  },
+  ["outPoint", "depType"]
+);
+
+export const DeCellDepVec = vector(CellDep);
```
Comment (PainterPuppets) [inline packages/codec/src/blockchain.ts]: I don't know if I'm missing the point, maybe it's `CellDepVec`?

File: packages/debugger/src/executor.ts [hunk 23]
```diff
@@ -37,6 +37,7 @@ export class CKBDebugger implements Executor {
    */
   private saveTmpTxFile(txSkeleton: TransactionSkeletonType): string {
     const debuggerData = parseDebuggerData(txSkeleton, this.loader);
+    // console.log('debuggerData', JSON.stringify(debuggerData));
```
Comment (PainterPuppets) [inline packages/debugger/src/executor.ts]: maybe can remove

File: examples/secp256k1-multisig-transfer/index.ts (hunk 5 above)
Comment (zhangyouxin) [reply examples/secp256k1-multisig-transfer/index.ts:158]: This is a HexString here

```ts
export const WitnessArgs = WitnessArgsOf({
  lock: HexifyCodec,
  inputType: HexifyCodec,
  outputType: HexifyCodec,
});
```

File: packages/base/lib/utils.d.ts (hunk 6 above)
Comment (zhangyouxin) [reply packages/base/lib/utils.d.ts]: Yes it's auto-generated, revert this file

File: packages/base/lib/indexer.js (hunk 20 above)
Comment (zhangyouxin) [reply packages/base/lib/indexer.js:1]: yes, not necessary

File: packages/codec/src/blockchain.ts (hunk 22 above)
Comment (zhangyouxin) [reply packages/codec/src/blockchain.ts]: typo

File: packages/transaction-manager/package.json (hunk 19 above)
Comment (zhangyouxin) [reply packages/transaction-manager/package.json:34]: This is used in `src`, previously it was a phantom dep

File: packages/codec/src/molecule/layout.ts (hunk 17 above)
Comment (zhangyouxin) [reply packages/codec/src/molecule/layout.ts:300]: eslint complains

File: packages/codec/package.json (hunk 16 above)
Comment (zhangyouxin) [reply packages/codec/package.json:34]: used `validators` previously

File: packages/ckb-indexer/tests/cell_collector.unit.test.ts (hunk 14 above)
Comment (zhangyouxin) [reply packages/ckb-indexer/tests/cell_collector.unit.test.ts:88]: previously it was used to test `validators`, thought it was useless afterwards, so removed when removing `toolkit`

Comment (zhangyouxin) [reply packages/ckb-indexer/tests/cell_collector.unit.test.ts:88]: Added back
